lib/opal/simple_server.rb

```diff
@@ -20,66 +21,69 @@ class Opal::SimpleServer
   NotFound = Class.new(StandardError)
 
   def initialize(options = {})
+    builder = options.fetch(:builder, nil)
     @prefix = options.fetch(:prefix, 'assets')
     @main = options.fetch(:main, 'application')
+    @app_dir = options.fetch(:app_dir, 'app')
     @index_path = nil
+
     yield self if block_given?
-    freeze
-  end
 
-  attr_accessor :main, :index_path
+    @production = ENV['RACK_ENV'] == 'production'
+    @start_time = Time.now.to_i
 
-  # @deprecated
-  # It's here for compatibility with Opal::Sprockets::Server
-  def append_path(path)
-    Opal.deprecation "`#{self.class}#append_path` is deprecated, please use `Opal.append_path(path)` instead (called from: #{caller(1, 1).first})"
-    Opal.append_path path
+    app_call = proc do |env|
+      call_index
+    rescue NotFound => error
+      [404, {}, [error.to_s]]
+    end
+
+    yield self if block_given?
+    @handler = Opal::RackHandler.new(app_call, { prefix: @prefix, main: @main, builder: builder,
+                                                 hot_updates: hot_updates == false ? : false : true,
+                                                 hot_javascript: hot_javascript,
+                                                 hot_ruby: hot_ruby })
   end
 
-  def call(env)
-    case env['PATH_INFO']
-    when %r{\A/#{@prefix}/(.*)\.m?js\z}
-      path, _cache_invalidator = Regexp.last_match(1).split('?', 2)
-      call_js(path)
-    else call_index
-    end
-  rescue NotFound => error
-    [404, {}, [error.to_s]]
+  attr_accessor :main, :index_path, :hot_updates, :hot_ruby, :hot_javascript
+
+  attr_reader :app_dir
+
+  def app_dir=(d)
+    @app_dir = d
+    Opal.append_path @app_dir
```

Comment:
Global state? I know we used to rely on it...

Reply:
Well, hot updates use the compiler directly, and it has no path otherwise.
Ill switch to using a Builder processor